### PR TITLE
EVA-2861 - Writer to write SS splits

### PR DESCRIPTION
**Pull request by sundarvenkata-EBI**

### CODE REVIEW

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SubmittedVariantSplittingPolicy.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            this.submittedVariantEntity = submittedVariantEntity;
+            this.ssHash = ssHash;
+            this.rsID = rsID;
+        }
+
+        public SubmittedVariantEntity getSubmittedVarianEntity() {
+            return submittedVariantEntity;
+        }
+
+        public String getSsHash() {
```
Comment (apriltuesday) [inline eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SubmittedVariantSplittingPolicy.java:44]:
```suggestion
        public String getSSHash() {
```
just for consistency with the rest of the codebase

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    public void write(@Nonnull List<? extends SubmittedVariantEntity> submittedVariantEntities)
+            throws MongoBulkWriteException, AccessionCouldNotBeGeneratedException, InstantiationException,
+            IllegalAccessException {
+        // We do this to avoid tedious generic type signatures in the method calls
+        List<SubmittedVariantEntity> svesWithDuplicateID = new ArrayList<>(submittedVariantEntities);
+
+        // All the following steps are designed to be idempotent
+        registerSplitCandidates(svesWithDuplicateID);
+        processSplitCandidates(this.allSplitCandidatesForCurrentBatch);
+        removeSplitCandidates(this.allSplitCandidatesForCurrentBatch);
```
Comment (tcezard) [inline eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java:94]: I'm not sure I understand why it is necessary to write candidates to the DB to remove them again right after the processing of the batch is complete.

**Review by nitin-ebi — APPROVED** [review]

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+    private void writeSplitCandidatesForCurrentBatchToDB(Map<String, List<SubmittedVariantEntity>> idSVEMap,
+                                                         Set<String> idsForSplitCandidateOperationsToWrite) {
+        // Only new split candidates in the current batch should be written to the database
+        List<SubmittedVariantOperationEntity> newSplitCandidatesForCurrentBatch = new ArrayList<>();
+        for (String id: idsForSplitCandidateOperationsToWrite) {
+            SubmittedVariantOperationEntity splitCandidateOperation = new SubmittedVariantOperationEntity();
+            List<SubmittedVariantEntity> svesThatShareSameID = idSVEMap.get(id);
+            // Ensure that non-duplicate SS IDs (or) SS IDs with multimap or mismatched alleles are not processed
+            if (svesThatShareSameID.size() > 1) {
```
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java:260]: Until the loop is complete in the previous method (getIdSVEMapForSplitCandidateOperations), we won't know which of the values in the map (that is being created in that method) only has one ID..

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java:94]: For the purposes of restartability, the remediation script for production run will process any outstanding split candidates before split writer is invoked. This way, any partially processed set of split candidate operations can be re-processed before proceeding to writing the splits.

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SubmittedVariantSplittingPolicy.java (hunk 1 above)
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SubmittedVariantSplittingPolicy.java:44]: Will address..

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java (hunk 2 above)
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java:94]: See [comment](https://github.com/EBIvariation/eva-accession/pull/342#discussion_r880505952)

File: eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                if(svesThatShouldGetNewIDs.contains(splitCandidate.getSubmittedVarianEntity())) {
+                    svesToCreateWithNewIDs.put(splitCandidate.getSsHash(), splitCandidate.getSubmittedVarianEntity());
+                }
+            }
+        }
+        excludeSSWithAlreadyUpdatedIDs(svesToCreateWithNewIDs);
+        removeCurrentSSEntriesInDBForSplitCandidates(svesToCreateWithNewIDs.keySet());
+        List<GetOrCreateAccessionWrapper<ISubmittedVariant, String, Long>> newlyCreatedSVEs =
+                this.submittedVariantAccessioningService.getOrCreate(new ArrayList<>(svesToCreateWithNewIDs.values()));
+        this.metricCompute.addCount(ClusteringMetric.SUBMITTED_VARIANTS_SS_SPLIT, newlyCreatedSVEs.size());
```
[the remark this answers is not in the document]
Comment (sundarvenkata-EBI) [reply eva-accession-clustering/src/main/java/uk/ac/ebi/eva/accession/clustering/batch/io/SSSplitWriter.java:121]: Good catch! It looks like I have to manually invoke [saveMetricCountsInDB](https://github.com/EBIvariation/eva-metrics/blob/c50de3085387c47a1a5234a302b823590703fef6/src/main/java/uk/ac/ebi/eva/metrics/metric/BaseMetricCompute.java#L43)

**Pull request merged**